File: android/src/main/kotlin/kr/yhs/flutter_kakao_maps/controller/overlay/handler/DimScreenControllerHandler.kt
package kr.yhs.flutter_kakao_maps.controller.overlay.handler

import com.kakao.vectormap.shape.DotPoints
import com.kakao.vectormap.shape.MapPoints
import com.kakao.vectormap.shape.DimScreenManager
import com.kakao.vectormap.shape.DimScreenLayer
import com.kakao.vectormap.shape.DimScreenCover
import com.kakao.vectormap.shape.PolygonOptions
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.Arrays
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asBoolean
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asInt
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asMap
import kr.yhs.flutter_kakao_maps.converter.PrimitiveTypeConverter.asString
import kr.yhs.flutter_kakao_maps.converter.ShapeTypeConverter.asDotPoints
import kr.yhs.flutter_kakao_maps.converter.ShapeTypeConverter.asMapPoints
import kr.yhs.flutter_kakao_maps.converter.ShapeTypeConverter.asPolygonOption
import kr.yhs.flutter_kakao_maps.converter.ShapeTypeConverter.asPolygonStylesSet
import kr.yhs.flutter_kakao_maps.converter.ShapeTypeConverter.asPolylineOption
import kr.yhs.flutter_kakao_maps.converter.ShapeTypeConverter.asPolylineStylesSet
import kr.yhs.flutter_kakao_maps.converter.ShapeTypeConverter.asShapeLayerOption

interface DimScreenControllerHandler {
  val dimScreenManager: DimScreenManager?

  val dimScreenLayer: DimScreenLayer?
    get = dimScreenManager?.getDimScreenLayer()
 
  fun dimScreenHandle(call: MethodCall, result: MethodChannel.Result) {
    val arguments = call.arguments!!.asMap<Any?>()
    if (dimScreenManager == null) {
      throw NullPointerException("DimScreenManager is null.")
    }

    when (call.method) {
      "setColor" -> setDimColor(arguments["color"]!!.asInt(), result::success)
      "setVisible" -> setDimVisible(arguments["visible"]!!.asBoolean(), result::success)
      "setDimCorver" -> setDimCorver(arguments["cover"]!!.asString().let { value: String ->
          DimScreenCover.entries.first { it.name == value }
      }, result::success)
      "addHighlightPolygonShape" -> {}
      "removeHighlightPolygonShape" -> {}
    }
  }

  fun setDimColor(color: Int, onSuccess: (Any?) -> Unit);

  fun setDimVisible(visible: Boolean, onSuccess: (Any?) -> Unit);

  fun setDimCorver(cover: DimScreenCover, onSuccess: (Any?) -> Unit);

  fun addDimHighlightPolygonShape(shape: PolygonOptions, onSuccess: (String) -> Unit)

  fun removeDimHighlightPolygonShape(polygonId: String, onSuccess: (Any?) -> Unit);
}

File: ios/kakao_map_sdk/Sources/kakao_map_sdk/controller/overlay/handler/DimScreenControllerHandler.swift
import Flutter
import KakaoMapsSDK

protocol DimScreenControllerHandler {
    var dimScreenManager: DimScreen { get }

    func setDimColor(color: UIColor, onSuccess: (Any?) -> Void)

    func setDimVisible(visible: Bool, onSuccess: (Any?) -> Void)

    func setDimCover(cover: DimScreenCover, onSuccess: (Any?) -> Void)

    func addDimHighlightPolygonShape(option: PolygonShapeOptions, visible: Bool, onSuccess: (String?) -> Void)

    func addDimHighlightMapPolygonShape(option: MapPolygonShapeOptions, visible: Bool, onSuccess: (String?) -> Void)

    func removeDimHighlightPolygonShape(shapeId: String, onSuccess: (Any?) -> Void)

    func removeDimHighlightMapPolygonShape(shapeId: String, onSuccess: (Any?) -> Void)

    func changeShapeVisible(shape: Shape, visible: Bool, onSuccess: (Any?) -> Void)

    func changeMapPolygonShape(shape: MapPolygonShape, styleId: String, position: [MapPolygon], onSuccess: (Any?) -> Void)

    func changePolygonShape(shape: PolygonShape, styleId: String, position: [Polygon], onSuccess: (Any?) -> Void)
}

extension DimScreenControllerHandler {
    func dimScreenHandle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = castSafty(call.arguments, caster: asDict)

        let polygonId = castSafty(arguments?["polygonId"], caster: asString)
        let mapPolygonShape: MapPolygonShape? = polygonId.flatMap { key in
            dimScreenManager.getHighlightMapPolygonShape(shapeID: key)
        }
        let polygonShape: PolygonShape? = polygonId.flatMap { key in
            dimScreenManager.getHighlightPolygonShape(shapeID: key)
        }
        let shape: Shape? = mapPolygonShape ?? polygonShape

        func requirePolygon() -> Shape? {
            guard let shape else {
                result(missingNativeResource(method: call.method, resource: "dim-screen polygon", id: polygonId))
                return nil
            }
            return shape
        }

        switch call.method {
        case "setColor":
            let color = UIColor(value: asUInt(arguments!["color"]!))
            setDimColor(color: color, onSuccess: result)
        case "setVisible": setDimVisible(visible: asBool(arguments!["visible"]!), onSuccess: result)
        case "setDimCover":
            let coverValue = asInt(arguments!["cover"]!)
            guard let cover = DimScreenCover(rawValue: coverValue) else {
                result(FlutterError(
                    code: "INVALID_DIM_SCREEN_COVER",
                    message: "Unknown DimScreenCover raw value: \(coverValue)",
                    details: nil
                ))
                return
            }
            setDimCover(cover: cover, onSuccess: result)
        case "addHighlightPolygonShape":
            let polygon = asDict(arguments!["polygon"]!)
            let position = asDict(polygon["position"]!)
            let positionType = asInt(position["type"]!)
            let visible = asBool(arguments!["visible"] ?? true)
            if positionType == 0 {
                let option = MapPolygonShapeOptions(payload: polygon)
                addDimHighlightMapPolygonShape(option: option, visible: visible, onSuccess: result)
            } else if positionType == 1 {
                let option = PolygonShapeOptions(payload: polygon)
                addDimHighlightPolygonShape(option: option, visible: visible, onSuccess: result)
            } else {
                result(FlutterMethodNotImplemented)
            }
        case "removeHighlightPolygonShape":
            guard requirePolygon() != nil, let id = polygonId else { return }
            if mapPolygonShape != nil {
                removeDimHighlightMapPolygonShape(shapeId: id, onSuccess: result)
            } else {
                removeDimHighlightPolygonShape(shapeId: id, onSuccess: result)
            }
        case "changePolygonVisible":
            guard let shape = requirePolygon() else { return }
            let visible = asBool(arguments!["visible"]!)
            changeShapeVisible(shape: shape, visible: visible, onSuccess: result)
        case "changePolygon":
            let styleId = asString(arguments!["styleId"]!)
            let rawPosition = asDict(arguments!["position"]!)
            let positionType = asInt(rawPosition["type"]!)
            if positionType == 0 {
                guard let mapPolygonShape else {
                    result(missingNativeResource(method: call.method, resource: "dim-screen map polygon", id: polygonId))
                    return
                }
                let position = MapPolygon(payload: rawPosition)
                changeMapPolygonShape(shape: mapPolygonShape, styleId: styleId, position: [position], onSuccess: result)
            } else if positionType == 1 {
                guard let polygonShape else {
                    result(missingNativeResource(method: call.method, resource: "dim-screen relative polygon", id: polygonId))
                    return
                }
                let position = Polygon(payload: rawPosition)
                changePolygonShape(shape: polygonShape, styleId: styleId, position: [position], onSuccess: result)
            } else {
                result(FlutterMethodNotImplemented)
            }
        default: result(FlutterMethodNotImplemented)
        }
    }
}
